### PR TITLE
SDKS-1710 WebAuthn Device Management

### PR DESCRIPTION
**Pull request by rodrigoareis**
# JIRA Ticket

[SDKS-1710](https://bugster.forgerock.org/jira/browse/SDKS-1710) WebAuthn Device Management

# Description

This change introduces a new method for removing WebAuthn keys from the server and locally in a single call.

Note: The ability to delete keys from the server relies on the availability of a new REST API for WebAuthn resources, which will be available on future versions of AM and IDC.

### CODE REVIEW

File: forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteWebAuthnRepository.kt [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+
+private val TAG = RemoteWebAuthnRepository::class.java.simpleName
+
+internal class RemoteWebAuthnRepository(val serverConfig: ServerConfig = Config.getInstance().serverConfig) :
+    WebAuthnRepository {
+
+    override suspend fun delete(publicKeyCredentialSource: PublicKeyCredentialSource) {
+        val credentialId = String(publicKeyCredentialSource.id)
```
Comment (witrisna) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteWebAuthnRepository.kt:34]: Should it be base64?

File: forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteWebAuthnRepository.kt [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        val userId = String(publicKeyCredentialSource.userHandle)
```
Comment (witrisna) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteWebAuthnRepository.kt:35]: Should it be base64?

File: forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteWebAuthnRepository.kt [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+        val findResponse = find(userId, credentialId)
+        val resourceId = findResponse.getString("uuid")
```
Comment (witrisna) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteWebAuthnRepository.kt:38]: getString will throw exception, may consider to use optString

File: forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteWebAuthnRepository.kt (hunk 1 above)
Comment (rodrigoareis) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteWebAuthnRepository.kt:34]: Looks like it's not required, the `toJson` and `fromJson` methods from the `PublicKeyCredentialSource` class handles the Base64 decode and decode.

File: forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteWebAuthnRepository.kt (hunk 2 above)
Comment (rodrigoareis) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteWebAuthnRepository.kt:35]: same as above

File: forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteWebAuthnRepository.kt (hunk 3 above)
Comment (rodrigoareis) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteWebAuthnRepository.kt:38]: good catch

**Review by witrisna — APPROVED** [review]

**Review by witrisna — DISMISSED** [review]

File: forgerock-auth/src/main/java/org/forgerock/android/auth/webauthn/FRWebAuthn.kt [hunk 4]
```diff
@@ -34,7 +50,26 @@ class FRWebAuthn(private val context: Context,
      * @param publicKeyCredentialSource The [PublicKeyCredentialSource] to be deleted
      */
     fun deleteCredentials(publicKeyCredentialSource: PublicKeyCredentialSource) {
-        webAuthnDataRepository.delete(publicKeyCredentialSource)
+        CoroutineScope(Dispatchers.IO).launch {
+            deleteCredentials(publicKeyCredentialSource, true)
```
[the remark this answers is not in the document]
Comment (spetrov) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/webauthn/FRWebAuthn.kt]: We've discussed this and decided to deprecate this method and preserve old behavior.

**Pull request merged**